Pattern.Neural.pde

```diff
@@ -0,0 +1,377 @@
+      gain_cols = splitTokens(gain_rows[i]);
+      for (int j=0; j < gain_cols.length; j++) {
+        gain[i][j] = float(gain_cols[j]);
+      }
+    }
+
+
+    n_sources = gain_cols.length;
+    n_nodes = lf_nodes.length;
+
+    //initialize
+    //k = 1/n_sources;
+    act = new float[n_sources][max_delay+2];
+    sensor_act = new float[n_nodes];
+
+    noise = new Random();
+    for (int i=0; i < n_sources; i++) {
+      for (int j=0; j < max_delay; j++) {
+        act[i][j]=(float)((noise.nextGaussian())*sigma.getValue()/100);
+      }
+    }
+
+    //start the sim
+    for (int t=0; t < max_delay; t++) {
+      step_simulation();
+    }
+  }
+
+  public void step_simulation() {
+
+    int t=max_delay;
+    for (int i=0; i<n_sources; i++) {
+      float w=0;
+      for (int j=0; j<n_sources; j++) {
+        w = w+C[j][i]*act[j][t-D[j][i]];
+      }
+      //floats can't possibly be helping at this point
+      act[i][t+1]=act[i][t]+tstep/.2*(-act[i][t]+(float)(k.getValue()/100/n_nodes)*w+(float)(sigma.getValue()/100*noise.nextGaussian()));
+    }
+    act[audio_source_left][t+1]=act[audio_source_left][t+1]+audio_in.left.get(0)*(float)(audio_wt.getValue()/1000);//*tstep;
+    act[audio_source_right][t+1]=act[audio_source_right][t+1]+audio_in.right.get(0)*(float)(audio_wt.getValue()/1000);//*tstep;
+    //System.out.println(act[84][t+1]);
+
+
+
+
+    //update node values
+    for (int i=0; i<n_nodes; i++) {
+      for (int j=0; j<n_sources; j++) {
+        sensor_act[i]=sensor_act[i] + gain[i][j]*act[j][t+1]*10;
+      }
+    }
+    //ugh
+    for (int j=1; j< max_delay+2; j++) {
+      for (int i=0; i<n_sources; i++) {
+        act[i][j-1]=act[i][j];
+      }
+    }
+  }
+
+  public void run(double deltaMs) {
+    audio_source_right = noise.nextInt(n_sources);
+    audio_source_left = noise.nextInt(n_sources);
+    for (int s=0; s < nsteps.getValue (); s++) {
+      step_simulation();
+    }
+    float dmin=min(sensor_act);
+    float dmax=max(sensor_act);
+    for (int i=0; i<n_nodes; i++) {
+      float v = (sensor_act[i]-dmin)/(dmax-dmin);
+      Node node = model.nodemap.get(lf_nodes[i]);
+      barlist = node.adjacent_bars();
+      for (Bar b : barlist) {
+        for (LXPoint p : b.points) {
+          colors[p.index]=lx.hsb((v*200-160+whatHue.getValuef())%360, 80, 80);
+          //colors[p.index]=palette.getColor(bv);
+        }
+      }
+    }
+  }
+}
+
+
+
+/** ********************************************************************
+ * Muse bandwidth energy pattern
+ *
+ * Does variation of Pixies pattern, with multiple layers
+ * @author Mike Pesavento
+ * original by Geoff Schmiddt
+ *
+ * Requires use of the muse_connect.pde file, which gives access to the muse headset data
+ * Also written by MJP. 
+ * Needs global object variable to be declared to access it in this pattern.
+ * Run "muse-io' from command prompt, eg
+ *    muse-io --preset 14 --osc osc.udp://localhost:5000
+ *
+ */
+class NeuroTracePattern extends LXPattern {
```

Comment:
It appears you took some old version of PixiePattern, rather than the GraphPattern.PixiePattern. I made some minor changes there, which you have lost here.

Reply:
Yeah, it was the original pattern from DBL, I havent looked at any updates you may have in the Mimsy version. Updated this multi-layer version to match the pattern in Pixies